src/NetAnalyzers/CSharp/Microsoft.CodeQuality.Analyzers/Maintainability/CSharpAvoidUninstantiatedInternalClasses.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeQuality.Analyzers.Maintainability;
+
+namespace Microsoft.CodeQuality.CSharp.Analyzers.Maintainability
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class CSharpAvoidUninstantiatedInternalClasses : AvoidUninstantiatedInternalClassesAnalyzer
+    {
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("MicrosoftCodeAnalysisPerformance", "RS1012:Start action has no registered actions.", Justification = "End action is registered in parent class.")]
+        public override void RegisterLanguageSpecificChecks(CompilationStartAnalysisContext startContext, ConcurrentDictionary<INamedTypeSymbol, object?> instantiatedTypes)
+        {
+            startContext.RegisterSyntaxNodeAction(context =>
+            {
+                var usingDirective = (UsingDirectiveSyntax)context.Node;
+
+                if (usingDirective.Alias != null &&
+                    context.SemanticModel.GetDeclaredSymbol(usingDirective) is IAliasSymbol aliasSymbol &&
```

Comment:
💡 It would be better to check for a generic type on the right hand side before calling `GetDeclaredSymbol`.

Reply:
I cannot directly check right side so I have used `Descendants` let me know if you were thinking about something else.

Reply:
Does this pattern cover it?

```csharp
var isGeneric = usingDirective.Name is
  GenericName
  or AliasQualifiedName { Name: GenericNameSyntax }
  or QualifiedNameSyntax { Right: GenericNameSyntax }
```

Reply:
I think it does! Do you want me to rewrite using this pattern instead of the call to descendants?

Reply:
Ping @sharwell

Reply:
I think using patterns is not a requirement/build enforcement, so you should be fine.